### PR TITLE
MSAL Automation Test Case Repair

### PR DESCRIPTION
**Pull request by fadidurah**
**What:** Fixing broken test cases in the msalautomationapp testpass. Also went through the  testpass and validated that all of the testcases match the ADO Test Plan instructions https://identitydivision.visualstudio.com/Engineering/_testPlans/execute?planId=1905195&suiteId=1905202.

**Why:** Fixing broken test cases and making sure they are all up to date with the test steps in ADO will lower the manual testing burden in future releases. Fixed cases can be used for reference in future automation cases.

**How:** Ran the entire msalautomationapp testpass and went through each case to find any differences or test failures. Ran the tests on the automation pipelines to insure significant increase in reliability and consistent passrates.

**Testing:** Ran the tests frequently on the ui automation pipelines, passrate and consistency has gone up dramatically. Pipelines:
https://dev.azure.com/IdentityDivision/Engineering/_build?definitionId=1743
https://dev.azure.com/IdentityDivision/Engineering/_build?definitionId=1742&_a=summary

### CODE REVIEW

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/local/TestCase99656.java [hunk 1]
```diff
@@ -40,13 +40,15 @@
...
 
+import org.junit.Ignore;
 import org.junit.Test;
 
 import java.util.Arrays;
 import java.util.concurrent.TimeUnit;
 
 // Interactive auth with force_login and step-up MFA
 // https://identitydivision.visualstudio.com/DefaultCollection/IDDP/_workitems/edit/99656
+@Ignore("https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1886086")
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/local/TestCase99656.java:50]: For this and any other AUTO_MFA test cases, Ryan was saying that we can try clicking on the "Call <Number>" button and see if that let's the flow proceed.

image

Can you try that out? If that works then we can remove the ignore annotation

Comment (fadidurah) [reply msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/local/TestCase99656.java:50]: I tried this, didn't seem to work. the called number doesn't responds and the sign in request doesn't finish. Same thing with text, but that's expected since I need to enter the code myself.

File: azure-pipelines/ui-automation/msal-broker-BrokerHost-nightly.yml [hunk 2]
```diff
@@ -183,7 +183,7 @@ stages:
                   --use-orchestrator \
                   --environment-variables "clearPackageData=true" \
                   --results-history-name "$(resultsHistoryName)" \
-                  --test-targets "package com.microsoft.identity.client.msal.automationapp.testpass.broker"
+                  --test-targets "package com.microsoft.identity.client.msal.automationapp.testpass.broker.brokerapi, package com.microsoft.identity.client.msal.automationapp.testpass.broker.atpop, package com.microsoft.identity.client.msal.automationapp.testpass.broker.crosscloud, package com.microsoft.identity.client.msal.automationapp.testpass.broker.flw"
```
Comment (shahzaibj) [inline azure-pipelines/ui-automation/msal-broker-BrokerHost-nightly.yml]: Is this actually needed? I think just having `com.microsoft.identity.client.msal.automationapp.testpass.broker` should've been good enough because the rest are just sub-packages.

If that is really not working then you may want to try:
`com.microsoft.identity.client.msal.automationapp.testpass.broker.*` and see if that works

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/crosscloud/TestCase1592465.java [hunk 3]
```diff
@@ -117,7 +122,17 @@ public void test_acquire_token_from_cross_cloud_after_acquiring_token_from_home_
...
+        final OnInteractionRequired crossCloudInteractionHandler = () -> {
+            final PromptHandlerParameters promptHandlerParameters =
+                    PromptHandlerParameters.builder()
+                            .prompt(PromptParameter.SELECT_ACCOUNT)
+                            .loginHint(userName)
+                            .broker(mBroker)
+                            .staySignedInPageExpected(GlobalConstants.IS_STAY_SIGN_IN_PAGE_EXPECTED)
+                            .build();
+            final AadPromptHandler promptHandler = new AadPromptHandler(promptHandlerParameters);
+            promptHandler.handlePrompt(userName, password);
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/crosscloud/TestCase1592465.java:136]: So this was previously passing without a prompt handler I guess.

Comment (shahzaibj) [reply msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/crosscloud/TestCase1592465.java:136]: I think the Expected Results in the Test Case in ADO are not clear. @iamgusain Can you review that test case in ADO ensure that Expected Results column clearly indicates whether to expect a prompt or not?

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/flw/TestCase833515.java [hunk 4]
```diff
@@ -104,57 +119,43 @@ public void test_833515() throws MsalException, InterruptedException, LabApiExce
...
-        chrome.launch();
-        chrome.navigateTo(MY_APPS_URL);
-        aadLoginComponentHandler.handleAccountPicker(username);
+//        aadLoginComponentHandler.handleAccountPicker(username2);
 
-        // we must see password prompt after sign out
-        final UiObject passwordField = UiAutomatorUtils.obtainUiObjectWithResourceId("i0118");
-        Assert.assertTrue(passwordField.exists());
+        // can sometimes take a while to actually be signed out
+        Thread.sleep(TimeUnit.SECONDS.toMillis(8));
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/flw/TestCase833515.java]: I believe the timeout was for when we were asserting logout in the browser. Since we removed that assertion, do we still need the sleep here?

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/local/TestCase99656.java (hunk 1 above)
Comment (shahzaibj) [reply msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/local/TestCase99656.java:50]: Probably need to follow up with Ryan on this

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/usgov/arlington/TestCase938368.java [hunk 5]
```diff
@@ -42,10 +48,13 @@
 import org.junit.Test;
 
 import java.util.Arrays;
+import java.util.concurrent.TimeUnit;
 
 // Interactive token acquisition with instance_aware=true, login hint present, and federated account,
 // and WW common authority
 // https://identitydivision.visualstudio.com/Engineering/_workitems/edit/938368
+// Adding a retry on failure, sometimes arlington login page fails to load
```
Comment (shahzaibj) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/usgov/arlington/TestCase938368.java:56]: Does it give an error when it fails to load? How often does this happen? 1 in 5 times? 1 in 10 times? 1 in 20 times? or some other frequency?

File: msalautomationapp/src/main/java/com/microsoft/identity/client/msal/automationapp/sdk/MsalSdk.java [hunk 6]
```diff
@@ -101,11 +103,17 @@ public MsalAuthResult acquireTokenInteractive(@NonNull MsalAuthTestParams authTe
             );
         }
 
-
         final AcquireTokenParameters acquireTokenParameters = acquireTokenParametersBuilder.build();
 
         pca.acquireToken(acquireTokenParameters);
 
+        // Adding small sleep to make sure acquireToken is up before we start handling UI
```
Comment (shahzaibj) [inline msalautomationapp/src/main/java/com/microsoft/identity/client/msal/automationapp/sdk/MsalSdk.java:110]: What happens otherwise?

**Review by shahzaibj — APPROVED** [review]
Minor comments, otherwise lgtm

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/atpop/TestCase1954183.java [hunk 7]
```diff
@@ -20,7 +20,7 @@
 //  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 //  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 //  THE SOFTWARE.
-package com.microsoft.identity.client.msal.automationapp.testpass.atpop;
+package com.microsoft.identity.client.msal.automationapp.testpass.broker.atpop;
```
Comment (somalaya) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/atpop/TestCase1954183.java]: Why is this file moved under broker folder?

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/crosscloud/TestCase1592510.java [hunk 8]
```diff
@@ -117,7 +121,16 @@ public void test_acquire_token_from_home_cloud_after_acquiring_token_from_cross_
         Assert.assertEquals("Verify Exception operation name", "authority", exception.getOperationName());
 
         // Acquire token interactively from home cloud, expected to get a different access token
-        final OnInteractionRequired homeCloudInteractionHandler = () -> {};
+        final OnInteractionRequired homeCloudInteractionHandler = () -> {
+            final PromptHandlerParameters promptHandlerParameters =
+                    PromptHandlerParameters.builder()
+                    .prompt(PromptParameter.SELECT_ACCOUNT)
```
Comment (somalaya) [inline msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/crosscloud/TestCase1592510.java:129]: Question : When we expect to see an interactive prompt - Are we also expecting to see a password prompt? Or just seeing the username field is enough in this case? Because setting the prompt type as SELECT_ACCOUNT will not autofill the password field but only autofills the username field.

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/crosscloud/TestCase1592465.java (hunk 3 above)
Comment (fadidurah) [reply msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/crosscloud/TestCase1592465.java:136]: This case and `1592510` are behaving differently on pipeline, the second request is not prompting for credentials. On my physical device, the second request always prompts for credentials. I'm probably just going to ignore these two for now and look at them later.

File: azure-pipelines/ui-automation/msal-broker-BrokerHost-nightly.yml (hunk 2 above)
Comment (fadidurah) [reply azure-pipelines/ui-automation/msal-broker-BrokerHost-nightly.yml]: This was added so that we don't attempt to run every test case and run into the 45m limit. The packages included here contain about half the test cases. I ended up adding a parameter to feed this through.

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/usgov/arlington/TestCase938368.java (hunk 5 above)
Comment (fadidurah) [reply msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/usgov/arlington/TestCase938368.java:56]: Sometimes gets the "This page isn't working" error. I'd say around 1 in 15 times it would happen. I also added code in this to refresh the page if the error shows up.

File: msalautomationapp/src/main/java/com/microsoft/identity/client/msal/automationapp/sdk/MsalSdk.java (hunk 6 above)
Comment (fadidurah) [reply msalautomationapp/src/main/java/com/microsoft/identity/client/msal/automationapp/sdk/MsalSdk.java:110]: I think I did this in response to some cases having a UiSelector failure where it wouldn't be able to locate the username or password box.

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/atpop/TestCase1954183.java (hunk 7 above)
Comment (fadidurah) [reply msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/atpop/TestCase1954183.java]: I think when I did some early repackaging, I thought all of these test were brokered so I moved them to a folder in broker, I can move 1954181 and 1954183 to msalonly and give them a package in that.

File: msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/crosscloud/TestCase1592510.java (hunk 8 above)
Comment (fadidurah) [reply msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/broker/crosscloud/TestCase1592510.java:129]: This test case is only showing the password prompt, with SELECT_ACCOUNT that prompt is being autofilled. The expected behavior for this case and 1592465 is a bit unclear, so i'm ignoring them for now.

**Review by somalaya — APPROVED** [review]
